torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -1330,6 +1309,12 @@ bool Reducer::rebuild_buckets() {
     return false;
   }
 
+  // Ensure reduction for previous backwards pass is finished. If user's model
+  // has unused parameters for example, this will raise an error recommending to
+  // run with find_unused_parameters=True, instead of the size mismatch exception
+  // below.
+  ensure_prior_reduction_finished();
```

Comment:
looks like `self.reducer._rebuild_buckets` is always invoked. Will it be sufficient if we move this to the front of the above if clause and then remove the other  `ensure_prior_reduction_finished` invocation in `prepare_for_backward`?

Reply:
Yes, I believe this should be okay. Originally I was concerned about `require_backward_grad_sync` flag, but I don't think its an issue. If we don't `require_backward_grad_sync` this iteration, we should still have the previous iteration reduction finished, where "previous iteration" here means the previous iteration that required grad sync, or the initial state of `require_finalize_`, which is false.